Zotero/Scenes/Detail/PDF/ViewModels/PDFReaderActionHandler.swift

```diff
@@ -1396,32 +1374,174 @@ final class PDFReaderActionHandler: ViewModelActionHandler, BackgroundDbProcessi
     /// Updates annotations based on insertions to PSPDFKit document.
     /// - parameter annotations: Annotations that were added to the document.
     /// - parameter viewModel: ViewModel.
-    private func add(annotations: [PSPDFKit.Annotation], selectFirst: Bool, in viewModel: ViewModel<PDFReaderActionHandler>) {
-        guard let boundingBoxConverter = self.delegate else { return }
+    private func add(annotations: [PSPDFKit.Annotation], in viewModel: ViewModel<PDFReaderActionHandler>) {
+        guard let boundingBoxConverter = delegate else { return }
 
         DDLogInfo("PDFReaderActionHandler: annotations added - \(annotations.map({ "\(type(of: $0));key=\($0.key ?? "nil");" }))")
 
-        let finalAnnotations = self.splitIfNeededAndProcess(annotations: annotations, state: viewModel.state)
+        let (keptAsIs, toRemove, toAdd) = transformIfNeeded(annotations: annotations, state: viewModel.state)
+        let finalAnnotations = keptAsIs + toAdd
+        for annotation in finalAnnotations {
+            if annotation.key == nil {
+                annotation.user = viewModel.state.displayName
+                annotation.customData = [AnnotationsConfig.keyKey: KeyGenerator.newKey]
```

Comment:
You shouldn't assign a new key to all "final" annotations. The ones that you re-do already have a key. So if you move it here I'd check for a key first and assign only to actually newly created ones.

Reply:
Isn't the clause `if annotation.key == nil` adequate?